### ISSUE TITLE
[BUG] LM converges much slower when gradient is accumulated in FP32.

### ISSUE THREAD

**Issue by torshie** [microsoft/DeepSpeed#5158]
**Describe the bug**
I'm using Megatron-LM (along with my own DeepSpeed integration), basic training works fine. Due to training stability issues, I enabled FP32 gradient accumulation, training loss no longer spikes, but it always much higher.

**To Reproduce**
I'm sorry, but I'm still looking for suggestion to isolate the bug, here is my training configuration:
* BF16 training
* DeepSpeedCPUAdam optimizer is used
* ZeRO-1, and set overlap_comm to True
* CPU offload is enabled
* tensor model parallel is set to 2
* gradient clipping 1.0
* llama-7b v1 model architecture
* 8GPU x 5 node training cluster
* tested on deepspeed 0.12 & 0.13
* if default gradient accumulation is used, loss is lower, training is unstable (unsurprising behavior), loss spikes quite often
* if FP32 gradient accumulation is used, loss is stable, but is much higher than expected
* in the screenshot below, v4 is trained with FP32 gradient accumulation while v3 is trained with default gradient accumulation

**Expected behavior**
with FP32 gradient accumulation, loss should be stable and not much higher or lower than BF16 gradient accumulation

**Screenshots**
![image](https://github.com/microsoft/DeepSpeed/assets/1214465/4780e35d-74cc-4bda-b5eb-7f1b8e3a4968)

I'm looking for suggestion to further isolate the bug,

**Comment by torshie** [comment]
I'm not using megatron-deepspeed because some features are missing

**Comment by torshie** [comment]
I did further tests, it can be reproduced in a single node with multiple GPU's. The key to reproduce is to enable data parallel, so the problem probably lies somewhere in the gradient reduction code.

**Comment by torshie** [comment]
After reading `stage_1_and_2.py` for a couple of times, I think the problem may be in the `backward()` method, I may not fully understand the source code, any comments are welcome:

* In certain setup, `DeepSpeedZeroOptimizer` may call `self.create_reduce_and_remove_grad_hooks()` to register a gradient reduction hook.
  * in my case set `overlap_comm` to `True` will trigger this call, I didn't realize this until now.
* When `loss.backward()` is called, the registered hook is called. 
* In the registered hook function, `reduce_independent_p_g_buckets_and_remove_grads()`  will be called when `is_gradient_accumulation_boundary` is `True`.
* Inside `reduce_independent_p_g_buckets_and_remove_grads()`, `get_gradient_for_reduction()` will be called to access the accumulated gradient attribute `.grad_accum`, which is updated by calling `fill_grad_accum_attribute()`.
* However, in the `backward()` method, `fill_grad_accum_attribute()` is called **after** `loss.backward()`.
* So basically when `is_gradient_accumulation_boundary` is `True`, `get_gradient_for_reduction()` will access the accumulated gradient **before** the final `fill_grad_accum_attribute()` call.

BTW, I think this is also the root cause of bug https://github.com/microsoft/DeepSpeed/issues/5174 which is easily reproducible with the provided source code.
